artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/ClientSessionFactoryImpl.java

```diff
@@ -632,11 +632,21 @@ private void failoverOrReconnect(final Object connectionID,
 
                connector = null;
 
-               reconnectSessions(oldConnection, reconnectAttempts, me);
+               boolean allSessionReconnected;
+               int failedReconnectSessionsCounter = 0;
+               do {
+                  allSessionReconnected = reconnectSessions(oldConnection, reconnectAttempts, me);
+                  if (oldConnection != null) {
+                     oldConnection.destroy();
+                  }
+                  oldConnection = connection;
 
-               if (oldConnection != null) {
-                  oldConnection.destroy();
+                  if (!allSessionReconnected) {
+                     failedReconnectSessionsCounter++;
```

Comment:
shouldn't you close the connection before setting it to null?


also... do you have a test showing the issue?


And 3rd... can you amend the commit with the JIRA?

```sh
git commit --amend
# on Vim or your configured editor, add the JIRA to the commit message (not the GitHub title)
git push origin -f
```

(If you happen to have a test meanwhile you're doing that?)

Reply:
1. the connection is not closed because the oldConnection is still using it. but the oldConnection might not get closed, I will fix that

2. quite tricky to come up with a test as described in the other comment

3. will do